pandas/core/internals/managers.py

```diff
@@ -1986,3 +1980,38 @@ def concatenate_block_managers(mgrs_indexers, axes, concat_axis, copy):
         blocks.append(b)
 
     return BlockManager(blocks, axes)
+
+
+def _discriminate_replace_and_coerce_block(
+    block, mask, new_rb, to_replace, value, inplace, convert, regex
+):
+    """
+    Process columns of block with values to be replaced separtely from rest of block.
+
+    This is done so that if, for example, a FloatBlock of shape (2, 2) is going to have
+    a value replaced from its first column (but not from its second column) such that
+    replacing that value will result in coercion to ObjectBlock, then only the first
+    column of the FloatBlock will get coerced.
+    """
+    if block.ndim > 1:
+        subblock_discriminator = mask.any(axis=1)
+        subblocks = [
+            block.getitem_block(subblock_discriminator),
+            block.getitem_block(~subblock_discriminator),
+        ]
+        subblock_masks = [mask[subblock_discriminator], mask[~subblock_discriminator]]
+    else:
+        subblocks = [block]
+        subblock_masks = [mask]
+
+    for subblock, subblock_mask in zip(subblocks, subblock_masks):
```

Comment:
any chance we can use Block.split_and_operate for some of this?

Reply:
@jbrockmendel thanks for the suggestion - just looked into it and it says
```
        f : callable accepting (1d-mask, 1d values, indexer)
```
while here we may be operating on multidimensional blocks, so it's no obvious to me how to use this here. Have marked this WIP while I update according to your other comments and think about this one